### ISSUE TITLE
Magnetic field in TrackParticle and ParticleBinning diagnostics

### ISSUE THREAD

**Issue by egelfer** [SmileiPIC/Smilei#717]
Hi! It seems that TrackParticle and ParticleBinning diagnostics output the magnetic field, which is half step shifted (both in space and time) from the particle. Is it possible to shift it back? Or maybe at least it would make sense to point it out in the documentation?

**Comment by mccoys** [comment]
Just a precision: the magnetic field is shifted by half a timestep compared to its value from Field diagnostics. Right? Is this what you meant?

**Comment by egelfer** [comment]
Not exactly. It is shifted by half a timestep and half a spatial step. In particular, if particle moves along x direction inside the plane wave, which is propagating in -x direction, then the vector of the transverse magnetic field (B_y,B_z) in TrackParticles/ParticleBinning diagnotics is rotated by the angle phi=timestep*(1+v_x)/2

**Comment by mccoys** [comment]
I don't understand how there could be a spatial shift. Is it due because of the wave displacement during that half timestep?

**Comment by egelfer** [comment]
I think it is due to the particle displacement during the half timestep, since there is the term timestep*v_x/2

**Comment by mccoys** [comment]
But what do you compare it with? The field diagnostic? If yes, i believe those are not properly aligned (by half a cell) in happi

**Comment by mccoys** [comment]
@egelfer Could you try your test using the option `keep_interpolated_fields`? It probably changes the results as the interpolated fields are calculated at the correct time with this option.

**Comment by egelfer** [comment]
Actually I used the option keep_interpolated_fields=["Ex","Ey","Ez","Bx","By","Bz"]. Do you want me to run the test without it?

**Comment by mccoys** [comment]
yes that would be good.

Now your result is puzzling. Those fields are supposed to be exactly those calculated for pushing particles

**Comment by egelfer** [comment]
Yes, I also assumed it and was very much surprized :)

**Comment by mccoys** [comment]
Ah well in fact I forgot that the magnetic field is interpolated at integer timesteps while the velocity is known at half-integer timesteps

![image](https://github.com/SmileiPIC/Smilei/assets/4188478/ba16f3ba-46b0-41ba-a144-91b814268828)

Thus the shift?

**Comment by egelfer** [comment]
Yes, I suppose that might be the source of the shift

**Comment by mccoys** [comment]
Did you manage to test without `keep_interpolated_fields`?

**Comment by mccoys** [comment]
Did you test without `keep_interpolated_fields`?

**Comment by mccoys** [comment]
I made some changes in the documentation in branch develop. Please reopen if necessary

**Comment by egelfer** [comment]
> Did you test without `keep_interpolated_fields`?

Sorry, I forgot to answer. In this case the simulation crushes. 

I also wanted to ask, is it possible to output the magnetic field in a way similar to when the force acting on a field is calculated (https://smileipic.github.io/Smilei/Understand/algorithms.html#the-pic-loop), i.e. (B_p^{(n+1/2)}+B_p^{(n-1/2)})/2 ? Unfortunately my method to recover the correct field is valid only for a plane wave, and it is not always the case.

**Comment by mccoys** [comment]
`keep_interpolated_fields` exactly retain the value of the fields that were obtained during interpolation. So you should have the force exactly.

**Comment by egelfer** [comment]
That's what I assumed! :) But it is not, what I've got. Here is the plot of average longitudinal momentum (darkred, obtained from the ParticleBinning diag) in the collision of electron-positron bunch with a pulse plane wave. Magenta curve corresponds to the integral of the (averaged over the particles) force F_x= [v x B]_x+E_x with the E and B fields obtained from TrackParticles diag. Orange curve corresponds to the integral of F_x'= [v x B']_x+E_x, where B' is B shifted by a half step. However, as I wrote above, it works only for a plane wave field.   

![image](https://github.com/SmileiPIC/Smilei/assets/34454093/2fd84aea-7024-4b49-8373-f3e1cc29be87)
